agent/agent.go

```diff
@@ -53,45 +53,61 @@ func main() {
 		os.Exit(1)
 	}
 
-	var previousCluster, previousInstanceID, previousContainerInstanceArn string
-	previousTaskEngine := engine.NewTaskEngine(cfg)
-	// previousState is used to verify that our current runtime configuration is
-	// compatible with our past configuration as reflected by our state-file
-	previousState, err := initializeStateManager(cfg, previousTaskEngine, &previousCluster, &previousContainerInstanceArn, &previousInstanceID)
-	if err != nil {
-		log.Crit("Error creating state manager", "err", err)
-		os.Exit(1)
-	}
+	var currentInstanceID, containerInstanceArn string
+	var taskEngine engine.TaskEngine
 
-	err = previousState.Load()
-	if err != nil {
-		log.Crit("Error loading previously saved state", "err", err)
-		os.Exit(1)
-	}
+	if cfg.Checkpoint {
+		var previousCluster, previousInstanceID, previousContainerInstanceArn string
+		previousTaskEngine := engine.NewTaskEngine(cfg)
+		// previousState is used to verify that our current runtime configuration is
+		// compatible with our past configuration as reflected by our state-file
+		previousState, err := initializeStateManager(cfg, previousTaskEngine, &previousCluster, &previousContainerInstanceArn, &previousInstanceID)
+		if err != nil {
+			log.Crit("Error creating state manager", "err", err)
+			os.Exit(1)
+		}
 
-	if cfg.Checkpoint && previousCluster != "" && previousCluster != cfg.Cluster {
-		log.Crit("Data mismatch; saved cluster does not match configured cluster. Perhaps you want to delete the configured checkpoint file?", "saved", previousCluster, "configured", cfg.Cluster)
-		os.Exit(1)
-	}
+		err = previousState.Load()
+		if err != nil {
+			log.Crit("Error loading previously saved state", "err", err)
+			os.Exit(1)
+		}
 
-	var currentInstanceID, containerInstanceArn string
-	var taskEngine engine.TaskEngine
-	if instanceIdentityDoc, err := ec2.GetInstanceIdentityDocument(); err == nil {
-		currentInstanceID = instanceIdentityDoc.InstanceId
-	} else {
-		log.Crit("Unable to access EC2 Metadata service to determine EC2 ID", "err", err)
-	}
+		if previousCluster != "" {
+			// TODO Handle default cluster in a sane and unified way across the codebase
+			configuredCluster := cfg.Cluster
+			if configuredCluster == "" {
+				configuredCluster = config.DEFAULT_CLUSTER_NAME
+			}
+			if previousCluster != configuredCluster {
+				log.Crit("Data mismatch; saved cluster does not match configured cluster. Perhaps you want to delete the configured checkpoint file?", "saved", previousCluster, "configured", configuredCluster)
+				os.Exit(1)
+			}
+			cfg.Cluster = previousCluster
+			log.Info("Restored cluster", "cluster", cfg.Cluster)
+		}
+
+		if instanceIdentityDoc, err := ec2.GetInstanceIdentityDocument(); err == nil {
+			currentInstanceID = instanceIdentityDoc.InstanceId
+		} else {
+			log.Crit("Unable to access EC2 Metadata service to determine EC2 ID", "err", err)
```

Comment:
Possibly.  I left this unchanged from @euank's previous commit and I'd like his thoughts here.